src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,251 @@
+    
+    pswfcs::Vector{Vector{T}}             # Pseudo-atomic wavefunctions
+    pswfc_ang_moms::Vector{Int}           # Angular momenta of the pseudo-atomic wavefunctions
+    pswfc_occs::Vector{T}                 # Occupations of the pseudo-atomic wavefunctions
+
+    rhoatom::Vector{T}                    # Pseudo-atomic charge density
+
+    rvlocpzdr::Vector{T}                  # (r_i V_{loc}(r_i) + Z) dr_i
+    r2projsdr::Vector{Vector{Vector{T}}}  # r_j^2 β_{li}(r_j) dr_i
+    
+    identifier::String                    # String identifying the PSP
+end
+
+"""
+    PspUpf(Zion::Number, lmax::Number, rgrid::Vector, drgrid::Vector, vloc::Vector,
+           vloc_interp, projs::Vector{Vector{Vector}}, projs_interp, h::Vector{Matrix},
+           pswfcs::Vector{Vector}, pswfc_ang_moms::Vector, pswfc_occs::Vector, rhoatom::Vector,
+           rvlocpzdr::Vector, r2projsdr::Vector{Vector{Vector}}; identifier="")
+
+Construct a Unified Pseudopotential Format pseudopotential. Currently only
+norm-conserving potentials are supported.
+"""
+function PspUpf(Zion, lmax, rgrid::Vector{T}, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                pswfcs, pswfc_ang_moms, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr;
+                identifier="") where {T}
+    length(projs) == length(h) || error("Length of projs and h do not agree.")
+    length(r2projsdr) == length(h) || error("Length of rgrid * projs * dr and h do not agree.")
+    length(pswfcs) == length(pswfc_ang_moms) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "angular momenta do not agree.")
+    length(pswfcs) == length(pswfc_occs) || error("Length of pseudo wfcs and pseudo wfc" *
+                                                      "occupations do not agree.")
+
+    return PspUpf{T}(Zion, lmax, rgrid, drgrid, vloc, vloc_interp, projs, projs_interp, h,
+                     pswfcs, pswfc_ang_moms, pswfc_occs, rhoatom, rvlocpzdr, r2projsdr, identifier)
+end
+
+function parse_upf_file(path; identifier=path)
+    pseudo = load_upf(path)
+
+    unsupported = []
+    pseudo["header"]["core_correction"] && push!(unsupported, "nonlinear core correction")
+    pseudo["header"]["has_so"] && push!(unsupported, "spin-orbit coupling")
+    pseudo["header"]["pseudo_type"] == "SL" && push!(unsupported, "semilocal potential")
+    pseudo["header"]["pseudo_type"] == "US" && push!(unsupported, "ultrasoft")
+    pseudo["header"]["pseudo_type"] == "PAW" && push!(unsupported, "plane-augmented wave")
+    pseudo["header"]["has_gipaw"] && push!(unsupported, "gipaw reconstruction data")
+    pseudo["header"]["pseudo_type"] == "1/r" && push!(unsupported, "Coulomb")
+    length(unsupported) > 0 && error("Pseudopotential contains the following unsupported "
+                                        * "features/quantities: $unsupported")
+
+    Zion = Int(pseudo["header"]["z_valence"])           # Pseudo-ion charge
```

Comment:
alignment in the code would help

Reply:
I may have made a mistake somewhere, but this seems to increase the number of allocations ~5x and the amount of allocation ~2x. It's also a fair bit slower.